### PR TITLE
enhance: Decouple shard client manager from shard cache

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #37115
the old implementation update shard cache and shard client manager at same time, which causes lots of conor case due to concurrent issue without lock.

This PR decouple shard client manager from shard cache, so only shard cache will be updated if delegator changes. and make sure shard client manager will always return the right client, and create a new client if not exist. in case of client leak, shard client manager will purge client in async for every 10 minutes.

### CODE REVIEW

**Comment by weiliu1031** [comment]
rerun go-sdk

File: internal/proxy/meta_cache.go [hunk 1]
```diff
@@ -1042,23 +1030,50 @@ func parseShardLeaderList2QueryNode(shardsLeaders []*querypb.ShardLeadersList) m
...
-		for _, shardLeaders := range db {
+	for dbName, dbInfo := range m.collLeader {
+		for collectionName, shardLeaders := range dbInfo {
 			if collectionSet.Contain(shardLeaders.collectionID) {
-				shardLeaders.deprecated.Store(true)
+				delete(dbInfo, collectionName)
 			}
 		}
+		if len(dbInfo) == 0 {
+			delete(dbInfo, dbName)
```
Comment (czs007) [inline internal/proxy/meta_cache.go:1075]: wrong

File: internal/proxy/meta_cache.go [hunk 2]
```diff
@@ -1042,23 +1030,50 @@ func parseShardLeaderList2QueryNode(shardsLeaders []*querypb.ShardLeadersList) m
...
 	log.Info("clearing shard cache for collection", zap.String("collectionName", collectionName))
-	if shards, ok := m.getCollectionShardLeader(database, collectionName); ok {
-		shards.deprecated.Store(true)
+	m.leaderMut.Lock()
+	defer m.leaderMut.Unlock()
+	dbInfo, ok := m.collLeader[database]
+	if ok {
+		delete(dbInfo, collectionName)
+		if len(dbInfo) == 0 {
+			delete(dbInfo, database)
```
Comment (czs007) [inline internal/proxy/meta_cache.go:1039]: wrong? should be  delete(m.collLeader, database)?

File: internal/proxy/shard_client.go [hunk 3]
```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
...
+func (c *shardClientMgrImpl) PurgeClient() {
+	ticker := time.NewTicker(600 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.closeCh:
+			return
+		case <-ticker.C:
+			shardLocations := globalMetaCache.ListShardLocation()
+			for nodeID := range c.clients.data {
```
Comment (czs007) [inline internal/proxy/shard_client.go:232]: is it safe to access the c.clients.data without any lock?

File: internal/proxy/shard_client.go [hunk 4]
```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
...
-	client, ok := c.clients.data[nodeID]
-	c.clients.RUnlock()
-
-	if !ok {
-		return nil, fmt.Errorf("can not find client of node %d", nodeID)
+func (c *shardClientMgrImpl) ReleaseClient(nodeID int64) {
+	c.clients.Lock()
+	defer c.clients.Unlock()
+	if client, ok := c.clients.data[nodeID]; ok {
+		client.Release()
```
Comment (czs007) [inline internal/proxy/shard_client.go:248]: Why does the client object remain in c.clients.data even after calling client.Release() ?

File: internal/proxy/shard_client.go [hunk 5]
```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
...
+	if !ok {
+		c.clients.Lock()
+		// Check again after acquiring the lock
+		client, ok = c.clients.data[info.nodeID]
+		if !ok {
+			// Create a new client if it doesn't exist
+			newClient, err := newShardClient(info, c.clientCreator)
 			if err != nil {
-				return err
+				return nil, err
```
Comment (czs007) [inline internal/proxy/shard_client.go:212]: failed to unlock  when error occur.

File: internal/proxy/shard_client.go [hunk 6]
```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
...
+	ticker := time.NewTicker(600 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.closeCh:
+			return
+		case <-ticker.C:
+			shardLocations := globalMetaCache.ListShardLocation()
+			c.clients.Lock()
+			defer c.clients.Unlock()
```
Comment (czs007) [inline internal/proxy/shard_client.go:233]: wrong ?

File: internal/proxy/meta_cache.go [hunk 7]
```diff
@@ -1042,23 +1030,50 @@ func parseShardLeaderList2QueryNode(shardsLeaders []*querypb.ShardLeadersList) m
...
 	log.Info("clearing shard cache for collection", zap.String("collectionName", collectionName))
-	if shards, ok := m.getCollectionShardLeader(database, collectionName); ok {
-		shards.deprecated.Store(true)
+	m.leaderMut.Lock()
+	defer m.leaderMut.Unlock()
+	dbInfo, ok := m.collLeader[database]
+	if ok {
+		delete(dbInfo, collectionName)
+		if len(dbInfo) == 0 {
+			delete(m.collInfo, database)
```
Comment (czs007) [inline internal/proxy/meta_cache.go:1039]: wrong

File: internal/proxy/shard_client.go [hunk 8]
```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
...
 
-func (c *shardClientMgrImpl) GetClient(ctx context.Context, nodeID UniqueID) (types.QueryNodeClient, error) {
-	c.clients.RLock()
-	client, ok := c.clients.data[nodeID]
-	c.clients.RUnlock()
-
-	if !ok {
-		return nil, fmt.Errorf("can not find client of node %d", nodeID)
+func (c *shardClientMgrImpl) ReleaseClient(nodeID int64) {
+	c.clients.Lock()
```
Comment (czs007) [inline internal/proxy/shard_client.go:245]: Rock

File: internal/proxy/meta_cache.go [hunk 9]
```diff
@@ -1042,23 +1030,50 @@ func parseShardLeaderList2QueryNode(shardsLeaders []*querypb.ShardLeadersList) m
...
+		delete(dbInfo, collectionName)
+		if len(dbInfo) == 0 {
+			delete(m.collLeader, database)
+		}
 	}
 }
 
+// used for Garbage collection shard client
+func (m *MetaCache) ListShardLocation() map[int64]nodeInfo {
+	m.leaderMut.Lock()
```
Comment (czs007) [inline internal/proxy/meta_cache.go:1046]: RLock?

**Pull request merged**